Goliath basic mob

Jacquerel opened the pull request:
## About The Pull Request

Converts Goliaths to the basic mob framework and gives them some new moves because I can't leave things well enough alone.
I am planning on touching all the lavaland fauna and then maybe even the icebox ones if I haven't got bored. The Golaith is the first because it is iconic.

https://www.youtube.com/watch?v=JNcKvMwT4-Q
Here's me getting killed by one as a demonstration. Despite my poor performance I would contend that they aren't a _lot_ more dangerous, but they are a little more dangerous.

The chief difference here is that they have two new attacks which they will only use in response to being attacked.
If fired at from range, they will target the attacker with a line of tentacles (it doesn't track you, so is easily sidestepped).
If attacked in melee, they will surround _themselves_ with tentacles, on a longer cooldown.

Something else you may notice in this video: I discovered that basic mobs are actually _too smart_ to be Lavaland fauna.
Typically (unlike their old form) a mob on our new AI system is smart enough to attack someone _the moment they come into range_ rather than only checking on predictable ticks, which would make using the Crusher an essentially unviable prospect.
To counteract this, Goliaths now have a delayed attack component which gives you a visual warning and short duration to get out of range before they swing at you. I will probably put this on all mining fauna that get reworked, it wouldn't be a terrible thing to put on other mobs to be honest.

Other changes: The goliath stun is now a status effect with _buckles_ you to the tentacle as if grabbed, as well as its previous effects. While this seems purely worse, any nearby helpers can now help-click on you to instantly remove the debuff.
Experiencing the effect of a Lobstrosity Rush Gland makes you immune to being grabbed by tentacles and an implanted one will automatically trigger and free you if you are hit, and the explosive effect of Brimdust also causes the tentacle to retract (although you'd need to take damage for this to happen). Using the tools of the land, you can make these creatures less threatening.

The ability for a Goliath to chain-apply the ability has now also been reduced, it won't refresh its duration if you are hit when already buckled.

When not occupied hounding miners, Goliaths will intermittently dig up the asteroid sand and eat any worms that this produces.
I also made some new sprites for riding a Goliath because they've been broken since the Lavaland mob update and also kind of were ugly before then anyway:
![image](https://github.com/tgstation/tgstation/assets/7483112/90580403-d82f-4c29-b3e1-6c462e01edda)

Other code changes:
- I made an element which only lets an attached object move every x seconds. This is because Goliaths are far too slow to use the speed system (the glide just looks bugged as hell) but one thing I am invested in when converting these is to make sure that they share the same behaviour when player or AI controlled. This is disabled while you're riding them because it was interminably slow.
- The Goliath tentacle trail uses a supertype object now shared with the Meteor Heart which did something kind of similar.

## Why It's Good For The Game

It begins the process of moving one of our larger subsets of NPCs onto the newer framework for NPC behaviour.
It adds a little bit more life to an iconic but slightly uninteresting foe which mostly just walked at you slowly.
This PR contains a few components I expect to apply more widely to other mobs in the future.

## Changelog

:cl:
refactor: Goliaths now use the Basic Mob framework, please report any unusual behaviour.
add: Goliaths learned a couple of new attacks which they will use in self-defence.
balance: Help-clicking a miner grabbed by Goliath tentacles will immediately free them, as will the effect of several items you can scavenge from around Lavaland.
image: New sprites for the Goliath saddle.
/:cl:

Jacquerel committed:
Bibles and Cowboy Hats can block bullets

(cherry picked from commit 88d4fe9919c243693b4c87548aaec85f229ab012)

Jacquerel committed:
Adds a stupid element probably nothing else is going to use, fuck

Jacquerel committed:
Makes it work

Jacquerel committed:
Cool abilities!

Jacquerel committed:
Start putting the AI in place

Jacquerel committed:
Goliaths do a little digging if they get bored, maybe find a tasty worm

Jacquerel committed:
Hack in a movement cooldown

Jacquerel committed:
Merge branch 'master' into big-basic-bitch

Jacquerel committed:
this MINER is PISSING me off

Jacquerel committed:
Golems can dig up the floor with their bare hands

Jacquerel committed:
Merge branch 'shovel-hands' into big-basic-bitch

Jacquerel committed:
Elementises mining mob resistance to being attacked from off screen

Jacquerel committed:
Moves crusher loot element to supertype for the future

Jacquerel committed:
Merge branch 'mining-cheese' into big-basic-bitch

Jacquerel committed:
Component which forecasts mob attacks and gives you time to get out of the way

Jacquerel committed:
Adds the old variants

Jacquerel committed:
You can do WHAT to the goliath?

Jacquerel committed:
Goliaths will no longer become extremely fixated on mushrooms
Fixes goliath riding sprites

Jacquerel committed:
Merge branch 'master' into big-basic-bitch

Jacquerel committed:
Don't retaliate while dead

Jacquerel committed:
Actually this is just an ability flag
Remove debug value

Jacquerel committed:
Adds default value to element so it is usable by admins out of the box, if they want to make a player very sad

Jacquerel committed:
Rename path script

Jacquerel committed:
Fucking dme

zoomachina commented:
> * I made an element which only lets an attached object move every x seconds. This is because Goliaths are far too slow to use the speed system (the glide just looks bugged as hell) but one thing I am invested in when converting these is to make sure that they share the same behaviour when player or AI controlled. This means they make pretty poor mounts but that was already true, they're a status symbol.
Does this mean riding a goliath is slower now? Is it slower than a wheelchair?

Jacquerel commented:
> Does this mean riding a goliath is slower now? Is it slower than a wheelchair?

Currently yes, I'm a little bit dissatisfied by that so I will see if I can come up with a solution.

CRITAWAKETS commented on code/datums/components/basic_mob_attack_telegraph.dm:
```diff
@@ -0,0 +1,82 @@
+/datum/component/basic_mob_attack_telegraph/Initialize(
+	telegraph_icon = 'icons/mob/telegraphing/telegraph.dmi',
+	telegraph_state = ATTACK_EFFECT_BITE,
+	telegraph_duration = 0.3 SECONDS,
```
0.3 is exceptionally short but really just on the verge of being dodgeable, as seen in the video, considering these are kind of a common enemy it might be a good idea to put this on 0.35 or 0.4 seconds, i dunno test and see what seems better

Jacquerel replied:
My experience with testing this is that 0.3 is about right for the existing crusher gameplay (frankly, "using melee anywhere" gameplay) of ducking in, clicking, and immediately back out of range. Quite possibly other mobs would be more lenient but to me it feels right for this one. Can always be adjusted again if it turns out that I made them into killing machines, but I don't think I have and I would consider myself only of average skill at best 😅

Jacquerel commented:
> Does this mean riding a goliath is slower now? Is it slower than a wheelchair?

> Do they still do that thing where if they run into a window they knock it outta frame? Cause that was always a nightmare when making one a pet. Also slow Goliath mounts sounds kind of ass, really hope there's a better way

ok I made it faster

https://github.com/tgstation/tgstation/assets/7483112/110e6497-ae18-40a8-a27c-e2391e1a7e40

Also yes they still have extreme pushing strength. You may just have to deal with people getting mad that you unseated their windows.

Jacquerel committed:
Rideable goliaths are a bit faster
You can't whack tamed goliaths to get them to trigger attacks
Goliaths can be tagged by the crusher
Goliaths won't get stuck trying to grab the same guy repeatedly if their cooldown is low

Jacquerel committed:
Deletes old overlay sprite

Jacquerel committed:
Merge branch 'master' into big-basic-bitch

Sealed101 commented:
i think one of the problems with tamed simple goliaths was that they were sometimes move on their own despite having a rider, and sometimes they would move into an object like a wall or table
how would tamed goliaths retaliate when attacked now that they're basic?

san7890 commented on code/__DEFINES/traits.dm:
```diff
@@ -614,6 +614,9 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 #define TRAIT_WEATHER_IMMUNE "weather_immune" //Immune to ALL weather effects.
 
+/// Cannot be grabbed by goliath tentacles
+#define TRAIT_TENTACLE_IMMUNE "tentacle_immune "
```
```suggestion
#define TRAIT_TENTACLE_IMMUNE "tentacle_immune"
```

san7890 commented on code/datums/elements/move_cooldown.dm:
```diff
@@ -0,0 +1,43 @@
+		return // Can't really eliminate involuntary movement but we'll try
+	if (!(source.movement_type & FLYING) && !source.has_gravity())
+		return // Skip newtonian movement if it's not under control
+	next_move_cache[REF(source)] = world.time + move_delay;
```
```suggestion
	next_move_cache[REF(source)] = world.time + move_delay
```

Jacquerel replied:
SMH I leave them off at work and add them superfluously here

san7890 commented on code/datums/mapgen/CaveGenerator.dm:
```diff
@@ -51,7 +51,7 @@
 	. = ..()
 	if(!weighted_mob_spawn_list)
-		weighted_mob_spawn_list = list(/mob/living/simple_animal/hostile/asteroid/goldgrub = 1, /mob/living/simple_animal/hostile/asteroid/goliath = 5, /mob/living/simple_animal/hostile/asteroid/basilisk = 4, /mob/living/simple_animal/hostile/asteroid/hivelord = 3)
+		weighted_mob_spawn_list = list(/mob/living/simple_animal/hostile/asteroid/goldgrub = 1, /mob/living/basic/mining/goliath/ancient = 5, /mob/living/simple_animal/hostile/asteroid/basilisk = 4, /mob/living/simple_animal/hostile/asteroid/hivelord = 3)
```
multiline+alphabetize this? will be useful for future refactor maintainability

san7890 commented on code/datums/memory/_memory.dm:
```diff
@@ -269,7 +269,7 @@
 		/mob/living/simple_animal/crab,
 		/mob/living/simple_animal/hostile/asteroid/basilisk/watcher,
-		/mob/living/simple_animal/hostile/asteroid/goliath/beast,
+		/mob/living/basic/mining/goliath,
```
alphabetize please

san7890 commented on code/game/objects/effects/temporary_visuals/effect_trail.dm:
```diff
@@ -0,0 +1,45 @@
+	/// Speed at which we chase target
+	var/move_speed = 3
+	/// What are we chasing?
+	var/datum/weakref/target
+	/// Stop spawning if we have this many effects already
+	var/max_spawned = 20
+	/// Do we home in after we started moving?
+	var/homing = TRUE
+	/// Handles chasing the target
+	var/datum/move_loop/movement
+
+/obj/effect/temp_visual/effect_trail/Initialize(mapload, atom/target)
```
ehhh feels kinda weird to have both the `weakref` of the target and the target be the same name... i wouldn't necessarily mind it if we stored a `var/atom/target` on the `effect_trail` type but since it's a weakref I think it should be named differently. i can get that we do the whole `src.target = WEAKREF(target)` but it just feels awk to have those two things be named the same when they are vastly different (where one is a datum subtype, and the other is a potential supertype for physical objects)

Jacquerel replied:
Easier to just not make it a weakref actually, it's already registered to the deletion signal so I couldn't tell you why I made it one in the first place.

san7890 commented on code/game/objects/structures/spawner.dm:
```diff
@@ -65,7 +65,7 @@
 	icon = 'icons/mob/simple/lavaland/nest.dmi'
 	spawn_text = "crawls out of"
-	mob_types = list(/mob/living/simple_animal/hostile/asteroid/goldgrub, /mob/living/simple_animal/hostile/asteroid/goliath, /mob/living/simple_animal/hostile/asteroid/hivelord, /mob/living/simple_animal/hostile/asteroid/basilisk, /mob/living/basic/wumborian_fugu)
+	mob_types = list(/mob/living/simple_animal/hostile/asteroid/goldgrub, /mob/living/basic/mining/goliath/ancient, /mob/living/simple_animal/hostile/asteroid/hivelord, /mob/living/simple_animal/hostile/asteroid/basilisk, /mob/living/basic/wumborian_fugu)
```
ditto multiline and alphabetize

san7890 commented on code/modules/mob/living/basic/lavaland/goliath/goliath.dm:
```diff
@@ -0,0 +1,225 @@
+	AddComponent(/datum/component/basic_mob_attack_telegraph)
+	AddComponentFrom(INNATE_TRAIT, /datum/component/shovel_hands)
+	if (tameable)
+		AddComponent(/datum/component/tameable, food_types = list(/obj/item/food/grown/ash_flora), tame_chance = 10, bonus_tame_chance = 5, after_tame = CALLBACK(src, PROC_REF(tamed)))
```
this is a pretty long line, multiline?

san7890 commented on code/modules/mob/living/basic/lavaland/goliath/goliath.dm:
```diff
@@ -0,0 +1,225 @@
+	balloon_alert(user, "affixing saddle...")
+	if (!do_after(user, delay = 5.5 SECONDS, target = src))
+		return
+	user.visible_message(span_notice("[user] mounts [attacking_item] to [src], [p_they()] are ready for riding!"))
```
this could also be something like `balloon_alert("ready to ride")`? up to you though

Jacquerel replied:
Loses visibility to other people nearby but the sprite visibly changes so that's fine.

san7890 commented on code/modules/mob/living/basic/lavaland/goliath/goliath_actions.dm:
```diff
@@ -0,0 +1,83 @@
+	/// Furthest range we can activate ability at
+	var/max_range = 7
+
+/datum/action/cooldown/goliath_tentacles/Activate(atom/target)
+	target = get_turf(target)
+	if (get_dist(owner, target) > max_range)
+		return FALSE
+
+	. = ..()
```
is there really no proc like `PreActivate()` where this could be passed in? Pretty silly to have the parent call in the middle proc like this in modern code imo, but that's an issue to implement another time. just making note of it now

san7890 commented on code/modules/mob/living/basic/lavaland/goliath/goliath_ai.dm:
```diff
@@ -0,0 +1,116 @@
+/datum/ai_controller/basic_controller/goliath
+	blackboard = list(
+		BB_TARGETTING_DATUM = new /datum/targetting_datum/basic/allow_items/goliath(),
```
```suggestion
		BB_TARGETTING_DATUM = new /datum/targetting_datum/basic/allow_items/goliath,
```

san7890 commented on code/modules/mob/living/basic/lavaland/goliath/tentacle.dm:
```diff
@@ -0,0 +1,119 @@
+	for (var/mob/living/victim in loc)
+		if (victim.stat == DEAD || HAS_TRAIT(victim, TRAIT_TENTACLE_IMMUNE))
+			continue
+		visible_message(span_danger("[src] grabs hold of [victim]!"))
```
could this be `victim.balloon_alert("grabbed by tentacle!")` or something

Jacquerel replied:
While I am mostly balloon-pilled I think that combat alerts like these should probably remain in the chatlog. I could put a balloon alert _alongside_ it if you want.

san7890 replied:
that's fine.

san7890 commented on code/modules/mob/living/basic/lavaland/goliath/tentacle.dm:
```diff
@@ -0,0 +1,119 @@
+	. = ..()
+	RegisterSignal(owner, COMSIG_CARBON_PRE_MISC_HELP, PROC_REF(on_helped))
+	RegisterSignals(owner, list(SIGNAL_ADDTRAIT(TRAIT_TENTACLE_IMMUNE), COMSIG_BRIMDUST_EXPLOSION), PROC_REF(release))
+	RegisterSignal(tentacle, list(COMSIG_QDELETING, COMSIG_GOLIATH_TENTACLE_RETRACTING), PROC_REF(on_tentacle_left))
```
```suggestion
	RegisterSignals(tentacle, list(COMSIG_QDELETING, COMSIG_GOLIATH_TENTACLE_RETRACTING), PROC_REF(on_tentacle_left))
```

c'mon nephew you did it right on the line above this

san7890 commented on code/modules/mob/living/basic/lavaland/goliath/tentacle.dm:
```diff
@@ -0,0 +1,119 @@
+/datum/status_effect/incapacitating/stun/goliath_tentacled/on_remove()
+	. = ..()
+	UnregisterSignal(owner, list(COMSIG_CARBON_PRE_MISC_HELP, SIGNAL_ADDTRAIT(TRAIT_TENTACLE_IMMUNE), COMSIG_BRIMDUST_EXPLOSION))
+	if (!tentacle)
```
```suggestion
	if (isnull(tentacle))
```

me suggesting `isnull()` will continue until morale improves

san7890 commented on code/modules/mob/living/living.dm:
```diff
@@ -1404,7 +1404,7 @@
 				/mob/living/simple_animal/crab,
 				/mob/living/simple_animal/hostile/asteroid/basilisk/watcher,
-				/mob/living/simple_animal/hostile/asteroid/goliath/beast,
+				/mob/living/basic/mining/goliath,
```
alphabetize please

san7890 commented on code/modules/procedural_mapping/mapGenerators/asteroid.dm:
```diff
@@ -21,7 +21,7 @@
 	spawnableTurfs = list()
 	spawnableAtoms = list(/mob/living/simple_animal/hostile/asteroid/basilisk = 10, \
 		/mob/living/simple_animal/hostile/asteroid/hivelord = 10, \
-		/mob/living/simple_animal/hostile/asteroid/goliath = 10)
+		/mob/living/basic/mining/goliath/ancient = 10)
```
multiline this properly and alphabetize please

san7890 commented on code/datums/components/revenge_ability.dm:
```diff
@@ -0,0 +1,55 @@
+	var/distance = get_dist(ability_user, attacker)
+	if (distance < min_range || distance > max_range)
+		return
+	if (targetting && !targetting.can_attack(victim, attacker))
```
does `can_attack()` check faction? i couldn't find it anywhere, because it would be really silly if someone added this component to something like a fireball wizard basic mob and they accidentally get smacked by someone in their faciton and then whammo blammo the player dies due to instant murderization component

if it doesn't and this doesn't seem like the right place to do it, i think the component should handle it

Jacquerel replied:
The targetting datum of most basic mobs will check faction (and does in this case) unless it's explicitly a mob which doesn't care about factions. I put it on here specifically to stop people whacking their own mount with a lasso in order to spur it to make tentacles 😓

san7890 commented on code/datums/components/basic_mob_attack_telegraph.dm:
```diff
@@ -0,0 +1,82 @@
+	SIGNAL_HANDLER
+	if (!isliving(target))
+		return
+	if (HAS_TRAIT(source, TRAIT_BASIC_ATTACK_FORECAST))
```
```suggestion
	if (HAS_TRAIT_FROM(source, TRAIT_BASIC_ATTACK_FORECAST, REF(src)))
```

odd feeling that this would pass and do an early return for no reason in case this trait gets used in other things for some reason. might be an unwarranted concern but it feels useful.

san7890 reviewed (changes requested):
a boatload of nits but also some codey things that should get looked at

Jacquerel commented:
> i think one of the problems with tamed simple goliaths was that they were sometimes move on their own despite having a rider, and sometimes they would move into an object like a wall or table how would tamed goliaths retaliate when attacked now that they're basic?

They won't retaliate against mobs of their faction, but currently I think that only means "attacks from their rider".
I should probably make them become friendly to all crew when tamed given that was the old behaviour, then it won't retaliate against anyone.

lizardqueenlexi commented:
I sort of wonder if the attack delay is too short - from the video it looks like the actual attack happens so fast that reacting is almost impossible (especially since basic mobs are so quick to launch an attack compared to simples).

Jacquerel committed:
Merge branch 'master' into big-basic-bitch

Jacquerel committed:
Multilines & reorders & removes silly characters
Weakref > ref

Jacquerel committed:
Goliath multilining & ability refactor
Fixed ai bug which actually made them never attack people who were grabbed oops

Jacquerel committed:
Remaining misc fixes

Jacquerel committed:
Restore previous taming faction function

Jacquerel committed:
Fixes misaligned bracket

Jacquerel committed:
maybe I shouldn't commit so fast

san7890 commented on code/modules/procedural_mapping/mapGenerators/asteroid.dm:
```diff
@@ -19,9 +19,11 @@
 	spawnableTurfs = list()
-	spawnableAtoms = list(/mob/living/simple_animal/hostile/asteroid/basilisk = 10, \
+	spawnableAtoms = list(
+		/mob/living/basic/mining/goliath/ancient = 10,\
+		/mob/living/simple_animal/hostile/asteroid/basilisk = 10, \
 		/mob/living/simple_animal/hostile/asteroid/hivelord = 10, \
-		/mob/living/simple_animal/hostile/asteroid/goliath = 10)
+		)
```
we don't need the backslashes here, we may have needed them in older versions of DM but we just don't need them anymore. also your end parenthesis was off

```dm
	spawnableAtoms = list(
		/mob/living/basic/mining/goliath/ancient = 10,
		/mob/living/simple_animal/hostile/asteroid/basilisk = 10,
		/mob/living/simple_animal/hostile/asteroid/hivelord = 10,
	)
```

san7890 commented on code/game/objects/structures/spawner.dm:
```diff
@@ -65,7 +76,13 @@
 	spawn_text = "crawls out of"
-	mob_types = list(/mob/living/simple_animal/hostile/asteroid/goldgrub, /mob/living/simple_animal/hostile/asteroid/goliath, /mob/living/simple_animal/hostile/asteroid/hivelord, /mob/living/simple_animal/hostile/asteroid/basilisk, /mob/living/basic/wumborian_fugu)
+	mob_types = list(
+		/mob/living/basic/mining/goliath/ancient,
```
where's the alphabetization sauce

iamcooldan commented:
"Goliaths will intermittently dig up the asteroid sand and eat any worms that this produces." amazing bit of flavor for explaining the lavaland food chain, lavaland has a ton of worms and presumably some of the sand that comes with the worm gets integrated into the rocky plates on their hide

Jacquerel committed:
is this more alphabetical or LESS alphabetical? why don't the viewers at home decide
